### PR TITLE
Modified redeploy workflow

### LINKED ISSUE #35: Rework redeploy workflow
Instead of doing stop and deploy operations to redeploy a service, make a redeploy script that uses `docker-compose` to build and restart the service

### PR DESCRIPTION
**Pull request by Harsh14901**
- For stopped services deploy normally
- For running services do not generate nginx entries again
- Added a parameter "-x" in deploy.sh to indicate no-nginx

Fixes #35

### CODE REVIEW

File: src/controllers/deploy.go
Comment (imperfect-fourth) [inline src/controllers/deploy.go:116]: I think that kwargs is unnecessarily complicating the flow. keep it as it was before and handle the new args in their respective routines

File: src/controllers/redeploy.go
Comment (imperfect-fourth) [inline src/controllers/redeploy.go]: keep the redeploy routine separate from deploy. I understand that there will be quite a lot of duplication of code but we can abstract that into a separate controller later.

**Review by imperfect-fourth — CHANGES_REQUESTED** [review]

File: src/controllers/handlers.go
Comment (imperfect-fourth) [inline src/controllers/handlers.go]: Rename this to `deployAction` because we don't need to export this since the type is only being used by the `controllers` package.

File: src/controllers/deploy.go
Comment (imperfect-fourth) [inline src/controllers/deploy.go]: If a service is in `stopped` state then its nginx entry does not exist. Adding the `x` flag here would mean that no nginx entry will be present for the service if a `redeploy` action is called in `stopped` state.

Comment (imperfect-fourth) [reply src/controllers/deploy.go]: Also, in this current logic a redeploy will never happen in a `running` state which shouldn't be the case.

File: src/controllers/redeploy.go
Comment (imperfect-fourth) [inline src/controllers/redeploy.go]: isn't the `command` already being set to `redeploy` [here](https://github.com/devclub-iitd/DeployBot/blob/67124a9c35235aadd7a3db209706dc97f3e37cb2/src/controllers/handlers.go#L33)

File: src/controllers/redeploy.go
Comment (imperfect-fourth) [inline src/controllers/redeploy.go]: Instead of calling a deploy routine from redeploy.go, you should write an internal routine for redeploy. The routines are only slightly different right now but that might change in the future and making use of if statements can only complicate the flow. They can both call the same `deploy.sh` with different flags for now.

**Review by imperfect-fourth — CHANGES_REQUESTED** [review]
`docker-compose up -d` does not rebuild the image; you would have to use `--build` flag in case of `redeploy` action.

File: src/controllers/deploy.go
Comment (imperfect-fourth) [inline src/controllers/deploy.go]: remove `params` from function parameters.

Comment (imperfect-fourth) [reply src/controllers/deploy.go]: rename this to `internalDeploy`

File: src/controllers/redeploy.go
Comment (imperfect-fourth) [inline src/controllers/redeploy.go]: remove `params` from here too

File: src/controllers/redeploy.go
Comment (imperfect-fourth) [inline src/controllers/redeploy.go:28]: I am thinking about a case where the state would have empty values for these field. That would probably happen if we call redeploy on a repo that hasn't been deployed. It's a corner case which I don't think would get triggered if we are careful about the usage, but in any case can you make sure this is handled here(and anywhere else where the same issue can happen)

File: src/controllers/redeploy.go
Comment (imperfect-fourth) [inline src/controllers/redeploy.go]: This wouldn't work since deploy would also require `access`, `server` and `subdomain` values. So if you see any of these values as empty, just log an error and return.
Also, remove the `state.Status == "stopped"` condition since that can happen organically.

**Review by imperfect-fourth — CHANGES_REQUESTED** [review]

File: src/controllers/redeploy.go
Comment (Harsh14901) [reply src/controllers/redeploy.go]: The `access`, `server` and `subdomain` fields are set by the deploy method using the `ActionInstance` object passed to it ([internalDeploy](https://github.com/devclub-iitd/DeployBot/blob/9bea7720a158ac205f69249394dddf65f9608bfa/src/controllers/deploy.go#L71)) which is in turn generated by the data obtained from slack request ([history.NewAction](https://github.com/devclub-iitd/DeployBot/blob/9bea7720a158ac205f69249394dddf65f9608bfa/src/controllers/deploy.go#L19)). In fact the `deploy` method is the first one to set up these state parameters for the service as evident by the [history.setState](https://github.com/devclub-iitd/DeployBot/blob/9bea7720a158ac205f69249394dddf65f9608bfa/src/controllers/deploy.go#L75) call. 

Also, the check for a `stopped` state was specifically added to avoid adding the `-x` flag to the deploy script which skips nginx regeneration which is undesirable since a stopped service doesn't have a nginx entry. In terms of abstractions, I think it makes more sense to "deploy" a stopped service before "re-deploying" it. So even if the methods for `redeploy` and `deploy` were to change in the future the logical flow would ensure that a stopped service gets deployed before being redeployed.

Comment (imperfect-fourth) [reply src/controllers/redeploy.go]: If you look at this [code](https://github.com/devclub-iitd/DeployBot/blob/26c87aa936d980eef0e164fca54e6d2152beb93b/src/slack/config.go#L122-L144), you'll see that the redeploy request is not sent with any data for `access`, `server` and `subdomain` fields. The purpose of `redeploy` command is to rebuild and run a service by taking these values from its previous state. 

You're right that those fields are set by deploy method using the `ActionInstance` object, which takes its values from the slack request, but as I said above, in a `redeploy` request these values are not sent. So when you come across empty state values, there is no way to get the values for `access`, `server` and `subdomain` fields for deploying.

Comment (Harsh14901) [reply src/controllers/redeploy.go]: Got it, Thanks!

**Review by imperfect-fourth — APPROVED** [review]
@arpit-saxena fetch these changes on mainvm and test DeployBot. If everything works fine, do a squash merge.

File: src/controllers/deploy.go
Comment (arpit-saxena) [inline src/controllers/deploy.go]: Shouldn't we also post to the slack channel that an error occurred while writing the log file?

File: src/controllers/restart.go
Comment (arpit-saxena) [inline src/controllers/restart.go]: Should read "Regenerating nginx entry.." so it's explicit

**Review by arpit-saxena — CHANGES_REQUESTED** [review]
Other than minor nits: In `SetState` function in `history.go`, we fail to set the state when requested state is `deploying` and the subdomain is not available. Can you take a look at how this function would change with adding the `redeploying` state? I think it might not change, but would appreciate another set of eyes on it so we don't miss anything.

**Comment by Harsh14901** [comment]
> Other than minor nits: In `SetState` function in `history.go`, we fail to set the state when requested state is `deploying` and the subdomain is not available. Can you take a look at how this function would change with adding the `redeploying` state? I think it might not change, but would appreciate another set of eyes on it so we don't miss anything.

We had brainstormed over this before and came to the same conclusion that no other change should be required, as `redeploying` is a transient state and subdomain availability is not an issue since the same subdomain for the service is being used.

File: src/controllers/deploy.go
Comment (arpit-saxena) [inline src/controllers/deploy.go]: Just make the error message have the `actionLog` too (like https://github.com/devclub-iitd/DeployBot/blob/6b8ffa26276d5c54d5f8882f395c67445174ae75/src/controllers/deploy.go#L41) so there would some context.

**Comment by arpit-saxena** [comment]
After that last fix, squash merge this PR and redeploy DeployBot. Then rebase #45

**Pull request merged**
